### PR TITLE
Add initial version of the program

### PR DESCRIPTION
**Pull request by milang**
- Demonstrates use of wildcard-csproj files in Pandell projects (requires Visual Studio 2017)
- Uses the latest C# version, 7.2
- Implements 5 commands:
    - `NTRightsNet help`: shows usage information
    - `NTRightsNet version`: shows version of the program; latest is `1.0.0.7`
    - `NTRightsNet get userName`: shows rights assigned to `userName` or `groupName` (can also be domain account, `domain\userName` or `domain\groupName`). Does not show rights assigned transitively via groups that the user is member of. Requires elevation.
    - `NTRightsNet add userName right1;right2;...;rightN`: adds specified rights to the `userName` or `groupName` (can also be domain account). Requires elevation.
    - `NTRightsNet remove userName right1;right2;...;rightN`: removes specified rights from the `userName` or `groupName` (can also be domain account). Requires elevation.

### CODE REVIEW

File: NTRightsNet.csproj [hunk 1]
```diff
@@ -0,0 +1,35 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <LangVersion>7.2</LangVersion>
+    <TargetFramework>net40</TargetFramework>
```
Comment (101100) [inline NTRightsNet.csproj:5]: Could you target .NET Standard 2.0 here or does the use of native calls make that nonsensical?

File: Program.cs [hunk 2]
```diff
@@ -0,0 +1,545 @@
...
+        //--------------------------------------------------
+        /// <summary>
+        /// Adds rights for the specified user.
+        /// </summary>
+        [NotNull] private static Result<string> CommandAddRights([NotNull, ItemNotNull] string[] args)
+        {
+            // verify arguments
+            if (args.Length != 3)
+            {
+                return new Failure<string>($"Expected exactly 2 parameters, username and semicolonSeparatedRights\n\n{((Success<string>)Program.CommandUsage(args)).Value}");
```
Comment (101100) [inline Program.cs]: `((Success<string>)Program.CommandUsage(args)).Value` is used four times.  How about defining a constant `Usage` that `CommandUsage` returns so you can use it for these cases?

It would be nice if it was declared right next to the command dictionary to make keeping them in sync easier and more obvious.

File: Program.cs [hunk 3]
```diff
@@ -0,0 +1,545 @@
...
+                    foreach (var right in rights) // release marshaled right strings
+                    {
+                        Marshal.FreeHGlobal(right.Buffer);
+                    }
+                }
+
+                return (addStatus == 0
+                    ? (Result<string>)new Success<string>(removeAll
+                        ? "&darkGray;Specified rights were successfully removed"
+                        : "&darkGray;All rights were successfully removed")
```
Comment (101100) [inline Program.cs]: This messages are reversed here.

**Review by 101100 — CHANGES_REQUESTED** [review]
Done with round 1.  I love the wildcard project file.

File: NTRightsNet.csproj (hunk 1 above)
Comment (milang) [reply NTRightsNet.csproj:5]: For me .NET standard only makes sense for libraries, not executables.

Note that this project does _not_ use .NET Core in any way (see note below). It builds a .NET Framework executable and targets old-ish .NET 4.0 for the widest compatibility with clean OSes (i.e. no dependencies other than .NET Framework that comes pre-installed on all reasonably old Microsoft OSes).

I did try to build a self-contained .NET Core executable. ~60+ MB with zillion DLLs for "Hello world" program is just _stupid_ and unacceptable (contrast this with self-contained Go programs that are a single executable of a few megabytes).

_Note_: `dotnet build` in instructions might be confusing (it certainly was to me). It does not mean that the project is using .NET core per se. It is only using `dotnet` CLI tool (that happens to be part of .NET core distribution that Visual Studio 2017 installer happens to add to PATH). The `dotnet build` locates MSBuild 15.0 for us and triggers `Build` target on the default project file (`.sln` in our case).

In other words `dotnet build` is really just a fancier version of our current `tools/Run.cmd` or `plicli` MSBuild runners that should "just work" on all development machines with Visual Studio 2017.

File: Program.cs (hunk 2 above)
Comment (milang) [reply Program.cs]: Good idea, updated.

File: Program.cs (hunk 3 above)
Comment (milang) [reply Program.cs]: Un-reversed 😃 .

**Comment by milang** [comment]
@101100 addressed feedback, ready for next round.

**Review by 101100 — APPROVED** [review]
Looks good to me.  🍪 

I also pulled a local copy which built and seemed to perform as advertised.

File: NTRightsNet.csproj [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    <Optimize>true</Optimize>
+    <DefineConstants>CODE_ANALYSIS;TRACE</DefineConstants>
+    <RunCodeAnalysis>true</RunCodeAnalysis>
+    <CodeAnalysisRuleSet>Properties/CodeAnalysis.ruleset</CodeAnalysisRuleSet>
+    <CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="JetBrains.Annotations" Version="11.1.0" />
+  </ItemGroup>
+  <ItemGroup>
```
Comment (amaechler) [inline NTRightsNet.csproj:30]: Beautiful. :)

**Review by amaechler — APPROVED** [review]
Looks great. I tested the `dotnet` tooling and reviewed the infrastructure (project files).

What a relief it will be to get rid of all those file references.
